### ISSUE TITLE
torch.sum() CPU is much slower for bool/uint8 tensors than int32/float32 tensors

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#37053]
## 🐛 Bug

Running single-core torch.sum() on a tensor of ints, bools, or uint8's is up to 16 times slower than a tensor of floats of the same length.

## To Reproduce

This issue might not occur on all processors, but it does show up on AMD Ryzen Threadripper 2970WX. Discussion on PR #33269 indicates that we don't see this issue on Intel Xeon 6148 or Intel Xeon 8180.

Running this script produces the following measurements: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/sum/sum_perf.py

device | data type | tensor size | torch.sum() time, seconds
-- | -- | -- | --
cpu | torch.float32 | 1000 | 0.000048
cpu | torch.float32 | 10000 | 0.000079
cpu | torch.float32 | 100000 | 0.000383
cpu | torch.float32 | 1000000 | 0.003521
cpu | torch.float32 | 10000000 | 0.037511
cpu | torch.int32 | 1000 | 0.000129
cpu | torch.int32 | 10000 | 0.000655
cpu | torch.int32 | 100000 | 0.005942
cpu | torch.int32 | 1000000 | 0.059008
cpu | torch.int32 | 10000000 | 0.610154
cpu | torch.bool | 1000 | 0.000129
cpu | torch.bool | 10000 | 0.000655
cpu | torch.bool | 100000 | 0.005943
cpu | torch.bool | 1000000 | 0.059016
cpu | torch.bool | 10000000 | 0.610713
cpu | torch.uint8 | 1000 | 0.000127
cpu | torch.uint8 | 10000 | 0.000642
cpu | torch.uint8 | 100000 | 0.005802
cpu | torch.uint8 | 1000000 | 0.05755
cpu | torch.uint8 | 10000000 | 0.596991

## Expected behavior

Summing ints should not be 16 times slower than summing floats. If anything, summing ints should take a little bit less time than summing floats.

## Environment

CPU: AMD Ryzen Threadripper 2970WX

PyTorch version: 1.6.0a0+851baf8
Is debug build: Yes
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.3 LTS
GCC version: (crosstool-NG 1.23.0.450-d54ae) 7.3.0
CMake version: version 3.16.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: TITAN RTX
GPU 1: TITAN RTX

Nvidia driver version: 440.33.01
cuDNN version: /usr/local/cuda-10.2.89/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy==1.17.5
[pip] torch==1.6.0a0+851baf8
[conda] magma-cuda101             2.5.1                         1    pytorch
[conda] mkl                       2019.5                      281    conda-forge
[conda] mkl-include               2019.5                      281    conda-forge
[conda] numpy                     1.17.5           py36h95a1406_0    conda-forge
[conda] torch                     1.6.0a0+851baf8           dev_0    <develop>

cc @VitalyFedyunin @ngimel

**Comment by kurtamohler** [comment]
I'm happy to look into this, since it seems like it might be specific to a subset of CPU architectures.

**Comment by ngimel** [comment]
fwiw, I'm seeing very similar relative performance on  Xeon(R) CPU E5-2698 v4
```
cpu torch.float32 1000 0.000003
cpu torch.float32 10000 0.000004
cpu torch.float32 100000 0.000016
cpu torch.float32 1000000 0.000121
cpu torch.float32 10000000 0.001461
cpu torch.int32 1000 0.000005
cpu torch.int32 10000 0.000012
cpu torch.int32 100000 0.000087
cpu torch.int32 1000000 0.000787
cpu torch.int32 10000000 0.054682
cpu torch.bool 1000 0.000006
cpu torch.bool 10000 0.000016
cpu torch.bool 100000 0.000102
cpu torch.bool 1000000 0.000792
cpu torch.bool 10000000 0.054013
cpu torch.uint8 1000 0.000005
cpu torch.uint8 10000 0.000013
cpu torch.uint8 100000 0.000086
cpu torch.uint8 1000000 0.000808
cpu torch.uint8 10000000 0.051542
```

**Comment by ngimel** [comment]
Ok, most of this is happening because 
a) benchmarking script actually benchmarks bool instead of int32 
b) bool and int32 and uint8 are promoted to int64 before reduction. 
When promoting to int32 which seems reasonable for most cases I get better results, still bool and uint8 are too slow, but that's probably because of preceding casting copies and not because of reduction itself:
```
pu torch.float32 1000 0.000003
cpu torch.float32 10000 0.000005
cpu torch.float32 100000 0.000018
cpu torch.float32 1000000 0.000119
cpu torch.float32 10000000 0.001387
cpu torch.int32 1000 0.000003
cpu torch.int32 10000 0.000004
cpu torch.int32 100000 0.000017
cpu torch.int32 1000000 0.000126
cpu torch.int32 10000000 0.001498
cpu torch.bool 1000 0.000008
cpu torch.bool 10000 0.000013
cpu torch.bool 100000 0.000074
cpu torch.bool 1000000 0.000574
cpu torch.bool 10000000 0.024307
cpu torch.uint8 1000 0.000006
cpu torch.uint8 10000 0.000009
cpu torch.uint8 100000 0.000052
cpu torch.uint8 1000000 0.000401
cpu torch.uint8 10000000 0.020086
```
fixed benchmark:
```
import torch
import time

torch.set_num_threads(1)

timed_iters = 100

for device in ['cpu']:
    for dtype in [torch.float, torch.int, torch.bool, torch.uint8]:
        out_dtype = torch.int if dtype != torch.float else torch.float
        for tensor_size in [1_000, 10_000, 100_000, 1_000_000, 10_000_000]:
            if dtype == torch.float:
                a = torch.rand(tensor_size, dtype=dtype, device=device)
            else:
                a = torch.randint(2, [tensor_size], dtype=dtype, device=device)
            for warmup_iter in range(2):
                start_time = time.time()
                for i in range(timed_iters):
                    a_sum = a.sum(dtype=out_dtype)
                # if device == 'cuda':
                #     torch.cuda.synchronize()
                total_time = time.time() - start_time

            time_per_iter = total_time / timed_iters

            print("%s %s %s %f" % (
                device,
                dtype,
                tensor_size,
                time_per_iter
            ))
```
Note, for masked_select one cannot unconditionally sum in int32, but it is safe at least for tensors that have less than 2**31 elements (don't know how overflow was handled previously in TH).

**Comment by vadimkantorov** [comment]
Even for sum reduction, if the underlying type is bool/uint8 and the reduced dim tensor size is sufficiently prudently small, the reduction can be done in int32 or even in a shorter type like int16, right? (modulo templated code bloat if it's a factor)

**Comment by kurtamohler** [comment]
Darn, sorry for the typo in my script. Thanks for investigating! I updated my repo with your edits, and here's the performance I get now:


device | data type | tensor size | torch.sum() time, seconds
-- | -- | -- | --
cpu | torch.float32 | 1000 | 0.000048
cpu | torch.float32 | 10000 | 0.000079
cpu | torch.float32 | 100000 | 0.00038
cpu | torch.float32 | 1000000 | 0.003478
cpu | torch.float32 | 10000000 | 0.03706
cpu | torch.int32 | 1000 | 0.000048
cpu | torch.int32 | 10000 | 0.000092
cpu | torch.int32 | 100000 | 0.000456
cpu | torch.int32 | 1000000 | 0.004184
cpu | torch.int32 | 10000000 | 0.041605
cpu | torch.bool | 1000 | 0.000126
cpu | torch.bool | 10000 | 0.000623
cpu | torch.bool | 100000 | 0.005558
cpu | torch.bool | 1000000 | 0.054987
cpu | torch.bool | 10000000 | 0.561507
cpu | torch.uint8 | 1000 | 0.000123
cpu | torch.uint8 | 10000 | 0.000607
cpu | torch.uint8 | 100000 | 0.005426
cpu | torch.uint8 | 1000000 | 0.053619
cpu | torch.uint8 | 10000000 | 0.5467

Since the real issue is with bools and uint8's I'll change the name of this issue.

I guess a quick fix could be to perform the sum with int32's if the bool/uint8 tensor is small enough that overflow would be impossible, otherwise promote to int64. The int32 sum could then be casted to int64 for consistency. Does that sound good to you? If so, I can go ahead and make that change.

**Comment by ngimel** [comment]
Yes, this sounds like a reasonable fix. Don't forget that uint8 tensors values can be up to 255, so you can safely sum only pretty small tensors. It probably makes sense to limit the fix to bool tensors only. Please don't change type promotion behavior (i.e. the output type by default should be int64, like it is now). Also can you please check if it will fix masked_select performance in #33269, at least for bool masks?

**Comment by kurtamohler** [comment]
I tried out the idea of summing a bool tensor into an int32 tensor when its size is small enough (less than or equal to 0x7fffffff), but that didn't speed up the summation significantly. So I looked at it with py-spy, and it turns out that with a 100000-element bool tensor, only about 7% of `sum()`'s runtime is spent inside the actual sum kernel. The other 93% is mostly within the `make_reduction` call, which creates the TensorIterator. This is true whether I'm using the original implementation that converts to 64-bit ints or if I'm using my experimental implementation that converts to 32-bit ints. Almost all of the time within `make_reduction` is spent in the `at::Tensor::to()` call that converts the bool tensor to an int tensor.

When I use a 100000-element int64 tensor as the input to `sum()`, `make_reduction` becomes a small portion of the runtime, because no conversion needs to happen, and the majority of runtime is spent inside the summation kernel.

So it doesn't matter whether we sum a bool tensor as int32 or int64, it's the conversion operation that takes the most time. I'll see if I can figure out why. Seems to me that  converting N bools to ints shouldn't have to take significantly longer than summing N ints.

**Comment by kurtamohler** [comment]
I made a script to measure the performance of `torch.to()` for several different pairs of input and output types: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/to/to_perf.py

With a 1 million-element tensor, on the AMD Threadripper, converting between bool/uint8/int32/int64 takes about 0.05 seconds in most cases--50 ns on average per element. Considering that this machine's frequency is limited between 2.2 - 3 GHz, that would be between 110 to 150 cycles for each conversion on average. That is extremely long. The latencies of a CPU's conversion instructions are usually on the order of 10's of cycles or less, and some of the latency will be hidden with pipelining and multiple lanes.

Running this on my Intel laptop, I see way better performance. The longest conversions in my benchmark (output to int64) took 0.000717 seconds for the full tensor, 0.717 ns per element, or about 2.4 cycles at 3.4 GHz on average.

I'll make a quick C++ benchmark to check out if the Threadripper just has ridiculously poor performance with conversions.

**Comment by kurtamohler** [comment]
I made a c++ int32-to-int64 performance measurement app: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/to/conversion_perf_standalone.cpp

The compiler args I used are in the makefile in the same directory.

I'm getting about 0.77 ns per conversion on my laptop, which is close to the pytorch performance. On the Threadripper, I get about 0.66 ns, so there doesn't seem to be anything inherently wrong with its mov 32-to-64 instruction, which is to be expected.

I'll keep looking for the bottleneck in pytorch.

**Comment by kurtamohler** [comment]
I just found out that the only reason I'm getting slow performance is because I always build pytorch with `DEBUG=1`. I tried building it without that flag and now both `at::to()` and `at::sum()` are much faster than my previous measurements.

device | data type | tensor size | torch.sum() time, seconds
-- | -- | -- | --
cpu | torch.float32 | 1000 | 0.000003
cpu | torch.float32 | 10000 | 0.000003
cpu | torch.float32 | 100000 | 0.000007
cpu | torch.float32 | 1000000 | 0.000053
cpu | torch.float32 | 10000000 | 0.003502
cpu | torch.int32 | 1000 | 0.000003
cpu | torch.int32 | 10000 | 0.000003
cpu | torch.int32 | 100000 | 0.000007
cpu | torch.int32 | 1000000 | 0.000055
cpu | torch.int32 | 10000000 | 0.003324
cpu | torch.bool | 1000 | 0.000005
cpu | torch.bool | 10000 | 0.000009
cpu | torch.bool | 100000 | 0.000051
cpu | torch.bool | 1000000 | 0.001208
cpu | torch.bool | 10000000 | 0.029895
cpu | torch.uint8 | 1000 | 0.000005
cpu | torch.uint8 | 10000 | 0.000007
cpu | torch.uint8 | 100000 | 0.000036
cpu | torch.uint8 | 1000000 | 0.000776
cpu | torch.uint8 | 10000000 | 0.028832
